File: packages/sdk-typescript/src/instrumentations/qdrant.ts
/**
 * Qdrant instrumentation for Tracelane.
 *
 * Wraps QdrantClient search() and upsert() to emit OTel spans.
 * Qdrant is Rust-native and Apache 2.0 — the natural vector DB companion
 * to Tracelane's stack. These spans enrich the agent trace with vector
 * retrieval context for RAG quality analysis.
 *
 * @example
 * ```ts
 * import { QdrantClient } from "@qdrant/js-client-rest";
 * import { instrumentQdrant } from "@tracelanedev/sdk/qdrant";
 *
 * const client = new QdrantClient({ url: "http://localhost:6333" });
 * instrumentQdrant(client);
 * const results = await client.search("my-collection", { vector: [...], limit: 5 });
 * ```
 */

import { SpanKind, SpanStatusCode, trace } from "@opentelemetry/api";

const tracer = trace.getTracer("@tracelanedev/sdk-qdrant", "0.1.0");

interface QdrantClientLike {
	search: (...args: unknown[]) => Promise<unknown>;
	upsert?: (...args: unknown[]) => Promise<unknown>;
	query?: (...args: unknown[]) => Promise<unknown>;
}

/**
 * Instrument a QdrantClient instance to emit OTel spans.
 *
 * @param client - A @qdrant/js-client-rest QdrantClient instance
 */
export function instrumentQdrant(client: QdrantClientLike): void {
	_patchSearch(client);
	if (client.upsert) {
		_patchUpsert(
			client as Required<Pick<QdrantClientLike, "upsert">> & QdrantClientLike,
		);
	}
	if (client.query) {
		_patchQuery(
			client as Required<Pick<QdrantClientLike, "query">> & QdrantClientLike,
		);
	}
}

function _patchSearch(client: QdrantClientLike): void {
	const originalSearch = client.search.bind(client);

	client.search = async (...args: unknown[]) => {
		const collectionName = String(args[0] ?? "unknown");
		const opts = args[1] as Record<string, unknown> | undefined;
		const limit = Number(opts?.limit ?? 10);

		return tracer.startActiveSpan(
			"qdrant.search",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "qdrant",
					"db.system": "qdrant",
					"db.operation.name": "search",
					"db.collection.name": collectionName,
					"qdrant.limit": limit,
				},
			},
			async (span) => {
				try {
					const results = (await originalSearch(...args)) as unknown[];
					const count = Array.isArray(results) ? results.length : 0;
					span.setAttribute("qdrant.results_count", count);
					span.setStatus({ code: SpanStatusCode.OK });
					return results;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

function _patchUpsert(client: {
	upsert: (...args: unknown[]) => Promise<unknown>;
}): void {
	const originalUpsert = client.upsert.bind(client);

	client.upsert = async (...args: unknown[]) => {
		const collectionName = String(args[0] ?? "unknown");
		const body = args[1] as Record<string, unknown> | undefined;
		const points = body?.points;
		const pointCount = Array.isArray(points) ? points.length : 0;

		return tracer.startActiveSpan(
			"qdrant.upsert",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "qdrant",
					"db.system": "qdrant",
					"db.operation.name": "upsert",
					"db.collection.name": collectionName,
					"qdrant.upsert_count": pointCount,
				},
			},
			async (span) => {
				try {
					const result = await originalUpsert(...args);
					span.setStatus({ code: SpanStatusCode.OK });
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

function _patchQuery(client: {
	query: (...args: unknown[]) => Promise<unknown>;
}): void {
	const originalQuery = client.query.bind(client);

	client.query = async (...args: unknown[]) => {
		const collectionName = String(args[0] ?? "unknown");

		return tracer.startActiveSpan(
			"qdrant.query",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "qdrant",
					"db.system": "qdrant",
					"db.operation.name": "query",
					"db.collection.name": collectionName,
				},
			},
			async (span) => {
				try {
					const result = (await originalQuery(...args)) as Record<
						string,
						unknown
					>;
					const points = result.points;
					if (Array.isArray(points)) {
						span.setAttribute("qdrant.results_count", points.length);
					}
					span.setStatus({ code: SpanStatusCode.OK });
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

File: packages/sdk-python/tracelane/instrumentations/qdrant.py
"""Qdrant instrumentation for Tracelane.

Wraps QdrantClient.search and QdrantClient.upsert to emit OTel spans.
Qdrant is Rust-native + Apache 2.0 — a natural companion to Tracelane's
stack. These spans enrich the full agent trace with vector retrieval context.

Example::

    from qdrant_client import QdrantClient
    from tracelane.instrumentations.qdrant import instrument_qdrant

    client = QdrantClient(url="http://localhost:6333")
    instrument_qdrant(client)
    results = client.search(collection_name="docs", query_vector=[...], limit=5)
    # Span emitted with qdrant.results_count and collection name
"""

from __future__ import annotations

from typing import Any

from opentelemetry import trace
from opentelemetry.trace import SpanKind, StatusCode

_tracer = trace.get_tracer("tracelane.qdrant", "0.1.0")


def instrument_qdrant(client: Any) -> None:
    """Wrap a QdrantClient instance to emit OTel spans.

    Instruments search(), upsert(), and delete() operations.

    Args:
        client: A qdrant_client.QdrantClient instance.

    Note:
        Vector values are never captured. Collection name, limit,
        and result count are recorded.
    """
    if hasattr(client, "search"):
        _patch_search(client)
    if hasattr(client, "upsert"):
        _patch_upsert(client)
    if hasattr(client, "query_points"):
        _patch_query_points(client)


def _patch_search(client: Any) -> None:
    original = client.search

    def _patched(*args: Any, **kwargs: Any) -> Any:
        collection = kwargs.get("collection_name", args[0] if args else "unknown")
        limit = kwargs.get("limit", 10)
        with _tracer.start_as_current_span(
            "qdrant.search",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "qdrant",
                "db.system": "qdrant",
                "db.operation.name": "search",
                "db.collection.name": str(collection),
                "qdrant.limit": int(limit),
            },
        ) as span:
            try:
                results = original(*args, **kwargs)
                count = len(results) if isinstance(results, list) else 0
                span.set_attribute("qdrant.results_count", count)
                span.set_status(StatusCode.OK)
                return results
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    client.search = _patched


def _patch_upsert(client: Any) -> None:
    original = client.upsert

    def _patched(*args: Any, **kwargs: Any) -> Any:
        collection = kwargs.get("collection_name", args[0] if args else "unknown")
        points = kwargs.get("points", args[1] if len(args) > 1 else [])
        point_count = len(points) if isinstance(points, list) else 0
        with _tracer.start_as_current_span(
            "qdrant.upsert",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "qdrant",
                "db.system": "qdrant",
                "db.operation.name": "upsert",
                "db.collection.name": str(collection),
                "qdrant.upsert_count": point_count,
            },
        ) as span:
            try:
                result = original(*args, **kwargs)
                span.set_status(StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    client.upsert = _patched


def _patch_query_points(client: Any) -> None:
    original = client.query_points

    def _patched(*args: Any, **kwargs: Any) -> Any:
        collection = kwargs.get("collection_name", args[0] if args else "unknown")
        limit = kwargs.get("limit", 10)
        with _tracer.start_as_current_span(
            "qdrant.query_points",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "qdrant",
                "db.system": "qdrant",
                "db.operation.name": "query_points",
                "db.collection.name": str(collection),
                "qdrant.limit": int(limit),
            },
        ) as span:
            try:
                result = original(*args, **kwargs)
                points = getattr(result, "points", None) or []
                span.set_attribute("qdrant.results_count", len(points))
                span.set_status(StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    client.query_points = _patched
